src/Pact/Eval.hs

```diff
@@ -222,35 +242,156 @@ loadModule m bod1 mi g0 = do
+evaluateConstraints
+  :: Module
+  -> HM.HashMap Text Ref
+  -> Info
+  -> Eval e (HM.HashMap Text Ref)
+evaluateConstraints Interface{} _ info =
+  evalError info $ "Impossible. interface found while appending meta-constraints to module"
+evaluateConstraints Module{..} evalMap info = foldMap (evaluateConstraint evalMap info) _mInterfaces
+  where
+    evaluateConstraint :: HM.HashMap Text Ref -> Info -> ModuleName -> Eval e (HM.HashMap Text Ref)
+    evaluateConstraint hm i ifn = do
+      -- load the interface refmaps via refstore
+      iRefs <- preview $ eeRefStore . rsModules . ix ifn . mdRefMap
+      case iRefs of
+        -- if nothing found, interface is not loaded, ergo not unfound
+        Nothing -> evalError info $
+          "Interface implemented in module, but not defined: <" ++ asString' ifn ++ ">"
+        Just iRefs' -> HM.foldrWithKey (solveConstraint i) (pure hm) iRefs'
+
+-- | Compare implemented member signatures and concatenate models
+--
+-- For each reference, we must check that it exists in the interface refmap,
+-- and if it does, we must check that it is both a def, and that we update
+-- any model information in the reference map with the concatenatation
+-- of the module models, as well as the interface models.
+solveConstraint
+  :: Info
+  -> Text
+  -> Ref
+  -> Eval e (HM.HashMap Text Ref)
+  -> Eval e (HM.HashMap Text Ref)
+solveConstraint info refName iref ehm = do
+  hm <- ehm
+  case HM.lookup refName hm of
+    Nothing -> pure hm
+    Just mref ->
+      case (iref, mref) of
+        -- only Direct if native
+        (Ref t, Ref s) ->
+          case (t, s) of
+            -- compare TDef only. This should never be comparing consts
+            -- since consts are stated in a module xor interface
+            (TDef _n _mn dt (FunType args rty) _ _ _,
+             TDef _n' _mn' dt' (FunType args' rty') _ _ _) -> do
+              -- compare deftype, return type, and if args lists are matching size, compare for correctness
+              when (dt /= dt') $ evalError info $ "deftypes mismatching: " ++ show dt ++ "\n" ++ show dt'
+              when (rty /= rty') $ evalError info $ "return types mismatching: " ++ show rty ++ "\n" ++ show rty'
+              when (length args /= length args') $ evalError info $ "mismatching argument lists: " ++ show args ++ "\n" ++ show args'
+              -- compare args by matching against name and type
+              forM_ (args `zip` args') $ \((Arg n ty _), (Arg n' ty' _)) -> do
+                when (n /= n') $ evalError info $ "mismatching argument names: " ++ show n ++ " and " ++ show n'
+                when (ty /= ty') $ evalError info $ "mismatching types: " ++ show ty ++ " and " ++ show ty'
+                return ()
+              pure hm
+            _ -> evalError info $ "found overlapping const refs - please resolve: " ++ show t 
+        _ -> evalError info $ "mismatching implementation signatures: \n" ++ show iref ++ "\n" ++ show mref
+      
```

Comment:
phew, that's a lot of code. oh well, the errors will be fantastic

Reply:
The errors have already been really helpful in debugging examples!